layer/tile/Google.js

```diff
@@ -24,35 +24,49 @@ L.Google = L.Class.extend({
 
 	// Possible types: SATELLITE, ROADMAP, HYBRID, TERRAIN
 	initialize: function (type, options) {
+		var _this = this;
+
+		this._ready = L.Google.isGoogleMapsReady();
+
 		L.Util.setOptions(this, options);
 
-		this._ready = google.maps.Map !== undefined;
-		if (!this._ready) L.Google.asyncWait.push(this);
+		this._googleApiPromise = L.Google.createGoogleApiPromise();
+
+		this._googleApiPromise
+		.then(function () {
+			_this._ready = true;
+			_this._initMapObject();
+			_this._update();
+		});
 
 		this._type = type || 'SATELLITE';
 	},
 
 	onAdd: function (map, insertAtTheBottom) {
-		this._map = map;
-		this._insertAtTheBottom = insertAtTheBottom;
+		var _this = this;
+		this._googleApiPromise
+		.then(function () {
+			_this._map = map;
+			_this._insertAtTheBottom = insertAtTheBottom;
 
-		// create a container div for tiles
-		this._initContainer();
-		this._initMapObject();
+			// create a container div for tiles
+			_this._initContainer();
+			_this._initMapObject();
 
-		// set up events
-		map.on('viewreset', this._reset, this);
+			// set up events
+			map.on('viewreset', _this._reset, _this);
 
-		this._limitedUpdate = L.Util.limitExecByInterval(this._update, 150, this);
-		map.on('move', this._update, this);
+			_this._limitedUpdate = L.Util.limitExecByInterval(_this._update, 150, _this);
```

Comment:
@brunob I noticed this is not really being used. I saw this was copied from L.TileLayer however they actually use it. 

I got into this because there is a noticeable lag to google maps due to it being proxied here. Can anything be improved for the lag? I think it is doubtful.